content/5_Plat/Bin_Jump.mdx

```diff
@@ -353,6 +353,216 @@ int main() {
 
 <IncompleteSection />
 
+### Solution (Company Queries II) 
+
+To find $\textrm{lca(a, b)}$, we can first jump the lower node of the two to the same depth as the other. Then, we lift both nodes up decrementally. At the end, the parent of either node is the LCA of the two.
+
+```cpp
+CodeSnip{Benq Template}
+    
+int N, Q, T=1;
+int depth[200005];
+int up[200005][20];
+vi adj[200005];
+
+void dfs(int v) {
+    FOR(i, 1, 20) {
+        up[v][i] = up[up[v][i-1]][i-1];
+    }
+
+    each(x, adj[v]) {
+        if(x != up[v][0]) {
+            depth[x] = depth[up[x][0]=v]+1;
+            dfs(x);
+        }
+    }
+}
+
+int jump(int x, int d) {
+	F0R(i, 20) {
+        if ((d>>i)&1) 
+            x = up[x][i];
+    }
+	return x; 
+}
+
+int LCA(int a, int b) {
+    if(depth[a] < depth[b]) swap(a, b);
+
+    a = jump(a, depth[a] - depth[b]);
+
+    if(a==b) return a;
+
+    R0F(i, 20) {
+        int aT = up[a][i]; int bT = up[b][i];
+        
+        if(aT != bT) {
+            a = aT; b = bT;
+        }
+    }
+
+    return up[a][0];
+}
+    
+int main() {
+    re(N, Q);
+    
+    F0R(i, N-1){
+        int b; re(b); --b;
+        adj[b].pb(i+1);
+    }
+
+    up[0][0] = 0, depth[0] = 0;
+    dfs(0);
+
+    F0R(i, Q) {
+        int a, b; re(a, b); --a; --b;
+        pr(LCA(a, b)+1,"\n");
+    }
+}
+```
+
+### Alternative Solution (Company Queries II)
```

Comment:
link the second tutorial (cp-algo) under this

Reply:
K